tests/charm++/atsync/atsync.C

```diff
@@ -0,0 +1,98 @@
+#include <stdio.h>
+#include <thread>
+
+#include "atsync.decl.h"
+
+/*readonly*/ CProxy_Main mainProxy;
+/*readonly*/ int nElements;
+/*readonly*/ int iters;
+
+/*mainchare*/
+class Main : public CBase_Main
+{
+public:
+  Main(CkArgMsg* m)
+  {
+    // Process command-line arguments
+    nElements = 5;
+    if (m->argc > 1)
+      nElements = atoi(m->argv[1]);
+    iters = 10;
+    if (m->argc > 2)
+      iters = atoi(m->argv[2]);
+    delete m;
+
+    // Start the computation
+    CkPrintf("Running atsync on %d processors for %d elements\n", CkNumPes(), nElements);
+
+    mainProxy = thisProxy;
+    CProxy_atsync arr = CProxy_atsync::ckNew();
+    for (int i = 0; i < nElements; i++)
+    {
+      arr[i].insert();
+    }
+    arr.doneInserting();
+
+    arr[0].start();
+  };
+
+  void done(void)
+  {
+    CkPrintf("All done\n");
+    CkExit();
+  };
+};
+
+class atsync : public CBase_atsync
+{
+private:
+  int counter;
+  int originalPE;
+
+public:
+  atsync()
+  {
+    CkPrintf("atsync (%d) created on %d\n", thisIndex, CkMyPe());
+    counter = 0;
+    originalPE = CkMyPe();
+    usesAtSync = true;
+  }
+
+  atsync(CkMigrateMessage* m) {}
+
+  void ResumeFromSync()
```

Comment:
The test doesn't really seem to "test" anything other than does it crash or hang. Should there be something maybe in ResumeFromSync to ensure that all elements are even actually hitting the AtSync. As written, the test would still pass even if ResumeFromSync is only ever called on 0.

Reply:
Yeah, that's all it really does, I originally wrote this to catch issues with `AtSync` where the internal state was getting messed up, but it was also useful in the unordered location updates issue, so failure is meant to be expressed as a hang (either deadlock or livelock). Adding more stuff is a good idea, though, I'll at the very least add a counter incremented in `ResumeFromSync` and then return that as a min reduction before exiting to make sure that all the objects are resumed the correct number of times.